### PR TITLE
Add Lyrics Tag support

### PR DESCRIPTION
**Pull request by broiler315**
Why not to add Lyrics tag support? I was waiting for this feature for several years and decided to do it myself. I simply ported Fauxdacious library "libfauxdtag" to original audacious branch and It Worked!!!

There is also need to port Fauxdacious lyricwiki plugin to display lyrics.
Thank You!

### CODE REVIEW

**Comment by jlindgren90** [comment]
It looks like this overwrites/reverts changes that have been made in Audacious since the code was forked. The port will need to be done more carefully than this. It's not just a matter of copying the files from Fauxdacious.

Also, please split up the changes into smaller commits rather than one mega-commit, with the changes explained in the commit messages.

**Comment by broiler315** [comment]
Ok, I've just cleaned the commit from any unnecessary changes. As a result there is only additon of reading & writing of 2 tags support (i.e. Lyrics & Publisher) in this commit. 
Now it's more simple to understand what I offered before, isn't it?

**Comment by radioactiveman** [comment]
Looks much better now at first glance, thanks. The JWT prefixes in comments are useless though in my opinion. I guess they are by Jim Turner (the Fauxdacious developer) to clarify it is changed/new code in his fork?

@jlindgren90, @kaniini:
General question: Why do we reinvent the wheel regarding tags? Why does Audacious not make use of TagLib? Their [website](https://taglib.org/) even lists Audacious as user :smile: Do you remember the reasons for reverting the code and writing libaudtag from scratch?

**Comment by jlindgren90** [comment]
We probably used taglib at some point in ancient history. I don't remember the exact issues with it but I think it was considered buggy. Chainsaw (Tony Vroon) was involved back then and I think mentored (?) the primary libaudtag developer at the time, though I also did a fair amount of work on it later.

I would be hesitant to switch out libaudtag for something else at this point because it's fairly well battle-tested by now.

**Comment by broiler315** [comment]
Yes, I removed Jim Turner's track in this case.

I think that the switch to TagLib is a good idea, because that project is developing and well supported at this time. It's used by VLC player and the Last.fm site due to it's description (however the Audacious is also in the list ....) and has more battle-tests and non-standard files support.
https://taglib.org/older.html

This is great tool for running on computers (c)
https://getmusicbee.com/forum/index.php?topic=36098.0

But on the other hand, it's a one more dependency and there is a need of making a big work for the migration. This work is worth further development of this project (e.g. other tags or formats support in future).

**Comment by kaniini** [comment]
the problem with taglib is that there are a lot of MP3 and Vorbis files out there which claim to have tags encoded as UTF-8 or ASCII, but are actually encoded with things like Shift-JIS or ISO8859-1.  this is why we added the character encoding detection feature, for example.  previously, taglib's assumptions about tag correctness is something we have had to fight, which is one of the main reasons why libaudtag was created, replacing taglib.

**Comment by kaniini** [comment]
(in general i would not be in favor of switching to taglib for these reasons)

File: .gitignore [hunk 1]
```diff
@@ -39,3 +39,4 @@ win32/override/README.txt
 *.patch
 *.so
 *~
+.idea
```
Comment (jlindgren90) [inline .gitignore:42]: Unrelated, please remove.

File: src/libaudcore/tuple.h [hunk 2]
```diff
@@ -105,6 +105,7 @@ class Tuple
         /* TODO: reorder these at next ABI break! */
         Description,   /* Track description */
         MusicBrainzID, /* MusicBrainz identifier */
+        Lyrics,        /* Lyrics from id3 tags. */
```
Comment (jlindgren90) [inline src/libaudcore/tuple.h:108]: Needs to be added at the end, after CatalogNum.

File: src/libaudtag/id3/id3-common.cc [hunk 3]
```diff
@@ -169,6 +169,24 @@ void id3_decode_comment (Tuple & tuple, const char * data, int size)
         tuple.set_str (Tuple::Comment, value);
 }
 
+void id3_decode_lyrics (Tuple & tuple, const char * data, int size)
```
Comment (jlindgren90) [inline src/libaudtag/id3/id3-common.cc:172]: Looks like an exact copy of `id3_decode_comment`, except for changing `Tuple::Comment` to `Tuple::Lyrics`. Please combine to avoid duplicating code.

File: src/libaudtag/id3/id3-common.cc [hunk 4]
```diff
@@ -327,7 +345,7 @@ static Index<char> id3_decode_pic_common (const char * data, int size, int encod
     int image_size = size - 1 - offset;
 
     AUDDBG ("Picture: type = %d, desc = %s, size = %d.\n", type,
-            (const char *) desc, image_size);
+           (const char *) desc, image_size);
```
Comment (jlindgren90) [inline src/libaudtag/id3/id3-common.cc:348]: Unnecessary formatting change, please revert.

File: src/libaudtag/id3/id3v24.cc [hunk 5]
```diff
@@ -519,6 +521,32 @@ static void add_comment_frame (const char * text, FrameDict & dict)
     g_free (utf16);
 }
 
+static void add_lyrics_frame (const char * text, FrameDict & dict)
```
Comment (jlindgren90) [inline src/libaudtag/id3/id3v24.cc:524]: Looks like an exact copy of `add_comment_frame`, except for changing `ID3_COMMENT` to `ID3_LYRICS`. Please combine the two to avoid duplicating code.

File: src/libaudtag/id3/id3v24.cc [hunk 6]
```diff
@@ -641,6 +672,7 @@ bool ID3v24TagModule::write_tag (VFSFile & f, const Tuple & tuple)
     int header_size, data_size, footer_size;
     bool syncsafe;
     int64_t offset;
+    bool mustpreservelength = false;  // Saving image or lyrics tags screw up the header length!
```
Comment (jlindgren90) [inline src/libaudtag/id3/id3v24.cc:675]: This doesn't look right. Please explain further.

File: src/libaudtag/id3/id3v24.cc [hunk 7]
```diff
@@ -664,6 +697,20 @@ bool ID3v24TagModule::write_tag (VFSFile & f, const Tuple & tuple)
     String comment = tuple.get_str (Tuple::Comment);
     add_comment_frame (comment, dict);
 
+    String lyrics = tuple.get_str (Tuple::Lyrics);
+    if (lyrics && lyrics[9])
```
Comment (jlindgren90) [inline src/libaudtag/id3/id3v24.cc:701]: What's the deal with `lyrics[9]`? Was that meant to be `lyrics[0]`?

File: src/libaudtag/id3/id3v24.cc [hunk 8]
```diff
@@ -664,6 +697,20 @@ bool ID3v24TagModule::write_tag (VFSFile & f, const Tuple & tuple)
...
+    {
+        add_lyrics_frame (lyrics, dict);
+        mustpreservelength = true;
+    }
+
+    if (mustpreservelength)
+    {
+        int decoder_length = tuple.get_int (Tuple::Length);
+        if (decoder_length > 0)
+            add_frameFromTupleInt (tuple, Tuple::Length, ID3_LENGTH, dict);
```
Comment (jlindgren90) [inline src/libaudtag/id3/id3v24.cc:711]: This is wrong. The `Tuple::Length` value usually comes from what the decoder calculated. We shouldn't be writing that value back to the ID3 tag.

**Review by jlindgren90 — CHANGES_REQUESTED** [review]

File: src/libaudtag/id3/id3v24.cc (hunk 7 above)
Comment (broiler315) [reply src/libaudtag/id3/id3v24.cc]: I think this wast just a "keypress" error. The "0" and "9" keys are close to each other, and unfortunately I coppied this error from that project.

File: src/libaudtag/id3/id3v24.cc (hunk 5 above)
Comment (broiler315) [reply src/libaudtag/id3/id3v24.cc]: Yes, I removed the duplicating code from this place (reading). And I did the same thing in the another place (writing). Please check out "id3_associate_memo" and "add_memo_frame" procedures.

File: src/libaudtag/id3/id3v24.cc (hunk 8 above)
Comment (broiler315) [reply src/libaudtag/id3/id3v24.cc]: Ok, I removed this. I didn't know before how this feature works and how the value is calculated. I haven't tested this part of code yet.

**Comment by jlindgren90** [comment]
Thanks for the updates, code is looking a lot better now 👍

File: .gitignore (hunk 1 above)
Comment (broiler315) [reply .gitignore]: Ok, I removed this. This was added to ignored list for cLion IDE. What IDE do you use to debug this project?

**Comment by broiler315** [comment]
I have already responsed to all your comments. I forced-pushed the changes, and you replied that the code looks better. What did I miss?

File: src/libaudcore/tuple.h (hunk 2 above)
Comment (broiler315) [reply src/libaudcore/tuple.h]: Done

File: src/libaudtag/id3/id3-common.cc (hunk 4 above)
Comment (broiler315) [reply src/libaudtag/id3/id3-common.cc]: Ok

File: src/libaudtag/id3/id3-common.cc (hunk 3 above)
Comment (broiler315) [reply src/libaudtag/id3/id3-common.cc]: Changed to 'id3_associate_memo'

**Review by jlindgren90 — CHANGES_REQUESTED** [review]

**Comment by jlindgren90** [comment]
> I have already responsed to all your comments. I forced-pushed the changes, and you replied that the code looks better. What did I miss?

@broiler315 Apparently I had "pending" comments that only I could see. Apologies for the confusion.

File: src/libaudtag/id3/id3v24.cc [hunk 9]
```diff
@@ -656,13 +661,23 @@ bool ID3v24TagModule::write_tag (VFSFile & f, const Tuple & tuple)
...
 
     String comment = tuple.get_str (Tuple::Comment);
-    add_comment_frame (comment, dict);
+    if (comment && comment[0])
+    {
+        add_memo_frame(ID3_COMMENT, comment, dict);
+    }
+
+    String lyrics = tuple.get_str (Tuple::Lyrics);
+    if (lyrics && lyrics[0])
```
[the remark this answers is not in the document]
Comment (broiler315) [reply src/libaudtag/id3/id3v24.cc]: It's the right thing, double check for empty text is excessive.

**Comment by broiler315** [comment]
No problem.

**Comment by jlindgren90** [comment]
Merged, thanks!
